casper/src/main/scala/coop/rchain/casper/MultiParentCasper.scala

```diff
@@ -59,182 +55,205 @@ object MultiParentCasper {
-        val currentBlockNumber  = maxBlockNum + 1
-        val earliestBlockNumber = currentBlockNumber - onChainState.shardConf.deployLifespan
-        for {
-          result <- DagOps
-                     .bfTraverseF[F, BlockMetadata](parentMetas)(
-                       b =>
-                         ProtoUtil
-                           .getParentMetadatasAboveBlockNumber(
-                             b,
-                             earliestBlockNumber,
-                             dag
-                           )
-                     )
-                     .foldLeftF(Set.empty[Signed[DeployData]]) { (deploys, blockMetadata) =>
-                       for {
-                         block        <- BlockStore[F].getUnsafe(blockMetadata.blockHash)
-                         blockDeploys = ProtoUtil.deploys(block).map(_.deploy)
-                       } yield deploys ++ blockDeploys
-                     }
-        } yield result
-      }
-      lfb <- dag.lastFinalizedBlockUnsafe
-    } yield CasperSnapshot(
-      dag,
-      lfb,
-      lca,
-      tips,
-      parents,
-      justifications,
-      deploysInScope,
-      maxBlockNum,
-      maxSeqNums,
-      onChainState
+
+      justifications <- parentHashes.toList.traverse(BlockDagStorage[F].lookupUnsafe(_))
+
+      // Calculate finalized fringe from justifications
+      msgMap    = dag.dagMessageState.msgMap
+      parents   = parentHashes.map(msgMap)
+      finalizer = Finalizer(dag.dagMessageState.msgMap)
+      // Get currently finalized bonds map
+      prevFringe       = finalizer.latestFringe(parents)
+      prevFringeHashes = prevFringe.map(_.id)
+      // Previous fringe state should be present (loaded from BlockMetadata store)
+      fringeRecord <- dag.fringeStates
+                       .get(prevFringeHashes)
+                       .liftTo {
+                         val fringeStr = PrettyPrinter.buildString(prevFringeHashes)
+                         val errMsg =
+                           s"Fringe state not available in state cache, fringe: $fringeStr"
+                         new Exception(errMsg)
+                       }
+
+      (prevFringeState, prevFringeRejectedDeploys) = fringeRecord
+      prevFringeStateHash                          = prevFringeState.bytes.toArray.toByteString
+      // TODO: for empty fringe bonds map should be loaded from bonds file (if validated in replay)
+      bondsMap <- if (prevFringe.isEmpty)
+                   justifications.head.bondsMap.pure[F]
+                 else
+                   RuntimeManager[F].computeBonds(prevFringeStateHash)
+
+      (_, newFringeOpt) = finalizer.calculateFinalization(parents, bondsMap)
+      newFringeHashes   = newFringeOpt.map(_.map(_.id))
+
+      prevFringeStr      = PrettyPrinter.buildString(prevFringeHashes)
+      prevFringeStateStr = PrettyPrinter.buildString(prevFringeState.toByteString)
+      _                  = println(s"PREV FRINGE $prevFringeStr")
```

Comment:
`Log` instead of `println`

Reply:
It's left over of debugging, I'll remove it.